language/stdlib/modules/validator_set.mvir

```diff
@@ -1,30 +1,93 @@
 module ValidatorSet {
+    import 0x0.LibraAccount;
     import 0x0.Vector;
 
-    struct TrustedPeer {
-        account_address: address,
+    struct ValidatorInfo {
+        addr: address,
+        network_signing_pubkey: bytearray,
+        network_identity_pubkey: bytearray,
+        consensus_pubkey: bytearray,
+        // TODO: voting power
+    }
+
+    // A current or prospective validator should publish one of these under their address
+    resource ValidatorConfig {
         network_signing_pubkey: bytearray,
         network_identity_pubkey: bytearray,
         consensus_pubkey: bytearray,
     }
 
     resource T {
-        validators: Vector.T<Self.TrustedPeer>,
+        validators: Vector.T<Self.ValidatorInfo>,
     }
 
-    make_new_validator_key(
-        account_address: address,
-        consensus_pubkey: bytearray,
+    // This can only be invoked by the special validator set address, and only a single time.
+    // Currently, it is invoked in the genesis transaction.
+    public initialize() {
+        // Only callable by the validator set address
+        assert(get_txn_sender() == 0x1D8, 1);
+
+        move_to_sender<T>(T { validators: Vector.empty<Self.ValidatorInfo>() });
+
+        return;
+    }
+
+    // Register the transaction sender as a candidate validator by creating a ValidatorConfig
+    // resource under their account
+    public register_candidate_validator(
         network_signing_pubkey: bytearray,
-        network_identity_pubkey: bytearray
-    ): Self.TrustedPeer {
-        let key: Self.TrustedPeer;
-        key = TrustedPeer {
-            account_address: move(account_address),
-            network_signing_pubkey: move(network_signing_pubkey),
-            network_identity_pubkey: move(network_identity_pubkey),
-            consensus_pubkey: move(consensus_pubkey),
-        };
-        return move(key);
+        network_identity_pubkey: bytearray,
+        consensus_pubkey: bytearray) {
+
+        move_to_sender<ValidatorConfig>(
+            ValidatorConfig {
+                network_signing_pubkey: move(network_signing_pubkey),
+                network_identity_pubkey: move(network_identity_pubkey),
+                consensus_pubkey: move(consensus_pubkey),
+            });
+
+        return;
     }
+
+    // Get the consensus public key for the validator candidate at the given address
+    public get_consensus_pubkey(addr: address): bytearray acquires ValidatorConfig {
+        let config_ref: &Self.ValidatorConfig;
+
+        config_ref = borrow_global<ValidatorConfig>(move(addr));
+        return *&move(config_ref).consensus_pubkey;
+    }
+
+    // Rotate a validator candidate's consensus public key. The change will not take effect until
+    // the next reconfiguration.
+    public rotate_consensus_pubkey(consensus_pubkey: bytearray) acquires ValidatorConfig {
+        let config_ref: &mut Self.ValidatorConfig;
```

Comment:
it's a little confusing to me which functions apply to each validator and which apply to the ValidatorSet